arroyo/processing/strategies/batching.py

```diff
@@ -233,3 +243,234 @@ def __flush(self) -> None:
         logger.debug("Offset commit took %dms", commit_duration)
 
         self.__batch = None
+
+
+ValuesBatch = Sequence[BaseValue[TPayload]]
+
+
+class BatchBuilder(Generic[TPayload]):
+    """
+    Accumulates Values in a Sequence of BaseValue.
+
+    It does not require offsets to be in monotonic order, though,
+    if offsets are not in monotonic order, it keeps track of the high
+    offset watermark as committable offset.
+    """
+
+    def __init__(
+        self,
+        max_batch_time: float,
+        max_batch_size: int,
+    ) -> None:
+        self.__max_batch_time = max_batch_time
+        self.__max_batch_size = max_batch_size
+        self.__values: MutableSequence[BaseValue[TPayload]] = []
+        self.__offsets: MutableMapping[Partition, Position] = {}
+        self.__init_time = time.time()
+
+    def append(self, value: BaseValue[TPayload]) -> None:
+        committables = value.committable
+        for partition, position in committables.items():
+            if (
+                partition not in self.__offsets
+                or self.__offsets[partition].offset <= position.offset
+            ):
+                self.__offsets[partition] = position
```

Comment:
Messages are in order, this can be a lot simpler.
```suggestion
        self.__offsets.update(value.committable)
```

Reply:
done